File: Multiple_Linear_Regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Feb 23 09:12:43 2017

@author: DK
"""
#importing packages
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#importing file
dataset = pd.read_csv("/Users/DK/Documents/Machine_Learning/Python/Macine_Learning_Projects/Multiple_Linear_Regression/50_Startups.csv")
X = dataset.iloc[:,:-1].values
Y = dataset.iloc[:,4].values

#categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
label_X = LabelEncoder()
X[:,3] = label_X.fit_transform(X[:,3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

#avoiding dummy variable trap
X = X[:,1:]

#training and test set
from sklearn.cross_validation import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2, random_state = 0)

#multiple linear regression model
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train,Y_train)

#prediction
y_pred = regressor.predict(X_test)

#backward elimination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50,1)).astype(int), values = X, axis = 1) #appending a constant columna t the beginning of the dataset as the statsmodel doesn't include any constant in the equation
X_opt = X[:,[0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()

#removing second column
X_opt = X[:,[0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()

#removing first column
X_opt = X[:,[0,3,4,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()

#removing fourth column
X_opt = X[:,[0,3,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()

File: Multiple_Linear_Regression.R
#insatlling packages
#install.packages("caTools")
#install_packages("ggplot2")

#importing libraries
library(caTools)

#importing dataset
dataset <- read.csv("/Users/DK/Documents/Machine_Learning/Python-and-R/Machine_Learning_Projects/Multiple_Linear_Regression/50_Startups.csv")

#categorical data
dataset$State<- factor(dataset$State, levels=c('New York','Florida','California'), labels=c(1,2,3))


#test and train sets
split <- sample.split(dataset$Profit,SplitRatio = 0.8)
training_set <- subset(dataset,split = TRUE)
test_set <- subset(dataset,split = FALSE)

#multiple linear regression model
mult_lin_reg <- lm(Profit~., data=training_set)
y_pred <- predict(mult_lin_reg,test_set)
summary(mult_lin_reg)

#back elimination
mult_lin_reg <- lm(Profit~.-State, data=training_set)
summary(mult_lin_reg)

mult_lin_reg <- lm(Profit~.-State-Administration, data=training_set)
summary(mult_lin_reg)

mult_lin_reg <- lm(Profit~R.D.Spend, data=training_set)
summary(mult_lin_reg)

